File: QytetetJava/src/modeloqytetet/Jugador.java

package modeloqytetet;

import java.util.ArrayList;

/**
 *
 * @author albertorodriguez
 */
class Jugador {
    
   //Atributo de la clase Sorpresa
    private Sorpresa cartaLibertad = null;
    
    //Atributo de la clase Casilla
    private Casilla casillaActual=null;
    
    //Atributo de TituloPropiedad
    private ArrayList<TituloPropiedad> propiedades = null; //new ArrayList();
    
    //Atributos propios 
    private boolean encarcelado = false;
    private String nombre;
    private int saldo = 7500;
    
    //--------------------------------------------------------------------------------------
    //-----------------------------------CONSTRUCTORES--------------------------------------
    
    //Se hace el constructor sin parametros privado para que no se pueda crear un jugador vacio
    
    public Jugador(String nombre){
        this.nombre = nombre;
        this.encarcelado = false;
        this.saldo = 7500;
        this.propiedades = null;
        this.cartaLibertad = null;
        this.casillaActual = new Casilla(0, 0, TipoCasilla.SALIDA);
    }
       
    //--------------------------------------------------------------------------------------
    //-----------------------------MODIFICADORES/CONSULTORES--------------------------------

    public Sorpresa getCartaLibertad() {
        return cartaLibertad;
    }

    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }

    public String getNombre() {
        return nombre;
    }

    public int getSaldo() {
        return saldo;
    }    
    
    public Casilla getCasillaActual() {
        return casillaActual;
    }

    public boolean getEncarcelado() {
        return encarcelado;
    }
    
    
    void setCasillaActual(Casilla casillaActual) {
        this.casillaActual = casillaActual;
    }

    void setEncarcelado(boolean encarcelado) {
        this.encarcelado = encarcelado;
    }
    
    void setCartaLibertad(Sorpresa carta){
        cartaLibertad = carta;
    }
    
    
    //--------------------------------------------------------------------------------------
    //-------------------------------------METODOS------------------------------------------
    
    public boolean tengoPropiedades(){
//        boolean tiene = false;
//        
//        for(TituloPropiedad p: propiedades)
//            if(p.getPropietario() == this)
//               tiene = true;
//        
//        return tiene;
        
        return (propiedades != null);
    }
    
    boolean actualizarPosicion(Casilla casilla){
        boolean tienePropietario;
        int costeAlquiler, coste;
        
        if(casilla.getNumeroCasilla() < casillaActual.getNumeroCasilla()){
            modificarSaldo(Qytetet.SALDO_SALIDA);
        }
        tienePropietario = false;
        setCasillaActual(casilla);
        
        if(casilla.soyEdificable()){
            tienePropietario = casilla.tengoPropietario();
            
            if(tienePropietario){
                this.encarcelado = casilla.propietarioEncarcelado();
                
                if(!(this.encarcelado)){
                    costeAlquiler = casilla.cobrarAlquiler();
                    modificarSaldo(-costeAlquiler);
                }
            }
        }
        else{
            if(casilla.getTipo() == TipoCasilla.IMPUESTO){
                coste = casilla.getCoste();
                modificarSaldo(coste);
            }
        }
        
        return tienePropietario;
    }
    
    boolean comprarTitulo(){
        boolean puedoComprar = false;
        boolean tengoPropietario;
        int costeCompra;
        TituloPropiedad titulo;
        
        if(casillaActual.soyEdificable()){
            tengoPropietario = casillaActual.tengoPropietario();
            if(!tengoPropietario){
                costeCompra = casillaActual.getCoste();
                if(costeCompra<=saldo){
                    titulo = casillaActual.asignarPropietario(this);
                    propiedades.add(titulo);
                    modificarSaldo(-costeCompra);
                    puedoComprar = true;
                }
            }
        }
        
        return puedoComprar;
    }
    
    Sorpresa devolverCartaLibertad(){
        Sorpresa s = cartaLibertad;
        
        if(cartaLibertad!=null){
            cartaLibertad = null;
        }
        else
            cartaLibertad = new Sorpresa("Hola", 0, TipoSorpresa.IRACASILLA);
        return cartaLibertad;
    }
    
    void irACarcel(Casilla casilla){
        setCasillaActual(casilla);
        setEncarcelado(true);
    }
    
    void modificarSaldo(int cantidad){
        if(cantidad > 0)
            saldo+=cantidad;
        else
            saldo -=cantidad;
    }
    
    int obtenerCapital(){
        int capitalTotal = saldo;
        int costeTotal = 0;
        int numCasasyHoteles = cuantasCasasHotelesTengo();
        
        for(TituloPropiedad p: propiedades){
            costeTotal += p.getCasilla().getCoste(); //Coste de la propiedad
            capitalTotal = costeTotal + (numCasasyHoteles * p.getCasilla().getPrecioEdificar());
            if(p.getHipotecada())
                capitalTotal -= p.getHipotecaBase();
        }
        
        return capitalTotal;
    }
    
    ArrayList<TituloPropiedad> obtenerPropiedadesHipotecadas(boolean hipotecada){
        ArrayList<TituloPropiedad> props = new ArrayList(); //o a null
        
        for(TituloPropiedad p: propiedades){
            if(p.getHipotecada() == hipotecada)
                props.add(p);                
        }
        
        return props;
    }
    
    void pagarCobrarPorCasaYHotel(int cantidad){
        int numeroTotal = cuantasCasasHotelesTengo();
        modificarSaldo(cantidad*numeroTotal);
    }
    
    boolean pagarLibertad(int cantidad){
        boolean tengo;
        tengo = tengoSaldo(cantidad);
        
        if(tengo){
            modificarSaldo(cantidad);
        }
        return tengo;
    }
    
    boolean puedoEdificarCasa(Casilla casilla){
        throw new UnsupportedOperationException("Sin Implementar"); 
    }
    
    boolean puedoEdificarHotel(Casilla casilla){
        throw new UnsupportedOperationException("Sin Implementar"); 
    }
    
    boolean puedoHipotecar(Casilla casilla){
        boolean esMia;
        esMia = esDeMiPropiedad(casilla);
        return esMia;
    }
    
    boolean puedoPagarHipoteca(Casilla casilla){
        throw new UnsupportedOperationException("Sin Implementar"); 
    }
    
    boolean puedoVenderPropiedad(Casilla casilla){
        boolean puedoVender = false;
        boolean esMia = esDeMiPropiedad(casilla);
        boolean hipotecada = casilla.estaHipotecada();
        
        if(esMia && !hipotecada)
            puedoVender = true;
  
        return puedoVender;
    }
    
    boolean tengoCartaLibertad(){
        boolean tengoLibertad = false;
        
        if(cartaLibertad != null)
            tengoLibertad = true;
        
        return tengoLibertad;
    }
    
    void venderPropiedad(Casilla casilla){
        int precioVenta = casilla.venderTitulo();
        modificarSaldo(precioVenta);
        eliminarDeMisPropiedades(casilla);
    }
    
    //DEVUELVE EL TOTAL DE CASAS Y HOTELES DE TODAS LA PROPIEDADES DEL JUGADOR.
    private int cuantasCasasHotelesTengo(){
        int total=0;
        
        for (TituloPropiedad p: propiedades){
           total += p.getCasilla().getNumCasas();
           total += p.getCasilla().getNumHoteles();
        }
        
        return total;
    }
    
    private void eliminarDeMisPropiedades(Casilla casilla){
        for(TituloPropiedad p: propiedades)
            if(p.getCasilla() == casilla)
                propiedades.remove(propiedades.indexOf(p.getCasilla()));
    }
    
    private boolean esDeMiPropiedad(Casilla casilla){
        boolean esPropiedad = false;
        
        for(TituloPropiedad p: propiedades)
            if(p.getCasilla().getTituloPropiedad() == casilla.getTituloPropiedad())
                esPropiedad = true;
        
        return esPropiedad;
    }
    
    private boolean tengoSaldo(int cantidad){
        return (saldo >= cantidad);
    }
    
    
    //----------------------------------------------------------------------
    //-----------------------------TO_STRING--------------------------------
    
    @Override
    public String toString(){
        return "\nJUGADOR:{ " + "\n   Nombre: " + nombre + 
                "\n   Saldo: " + Integer.toString(saldo) + 
                "\n   Encarcelado: " + encarcelado +
                "\n   Casilla Actual: " + casillaActual.toString() + 
                "\n   Carta Libertad: " + cartaLibertad +
                "\n   Propiedades: " + propiedades + "\n}";
        
        //AÑADIR A UN STRING LO QUE QUIERO IMPRIMIR
    }
    
    
}

File: QytetetRuby/pkg/QytetetRuby-0.0.1/lib/jugador.rb
#encoding: utf-8

module ModeloQytetet

class Jugador
    attr_accessor :encarcelado
    attr_accessor :nombre
    attr_accessor :saldo
    
    #Atributo de la clase sorpresa
    attr_accessor :carta_libertad
    
    #Atributo de la clase Casilla
    attr_accessor :casilla_actual
    
    #Atributo de la clase TituloPropiedad
    attr_accessor :propiedades
    
 
    
  def initialize(nombre)
    @encarcelado = false
    @nombre = nombre
    @saldo = 7500
    @propiedades = Array.new
    @casilla_actual = nil
    @carta_libertad = nil
  end
  
  #----------------------TO_S----------------------

   def to_s()
      "\n       Nombre: #{@nombre} 
       Encarcelado: #{@encarcelado} 
       Saldo: #{@saldo} 
       Carta Libertad: #{@cartaLibertad} 
       Casilla Actual: #{@casillaActual}
      "
    end
  
  #----------------------METODOS----------------------
  
   #DEVUELVE VERDADERI CUANDO EL JUGADOR ES PROPIETARIO DE ALGUN TITULO PROPIEDAD Y FALSO EN CASO CONTRARIO
  def tengo_propiedades
    tengo = false
    @propiedades.each do |p|
      if(this == p.propietario)
        tengo = true
      end
    end
    tengo
    
    #if(@propiedades != nil)
  end
  
  def actualizar_posicion(casilla)
    
  end
  
  def comprar_titulo
    
  end
  
  #Devuelve la carta Sorpresa cartaLibertad, pues el jugador ya ha hecho uso de ella. Esto implica que el jugador se queda sin esa carta.
  def devolver_carta_libertad
    
  end
  
  def ir_a_carcel(casilla)
   
  end
  
  #AÑADE AL SALDO LA CANTIDAD DEL ARGUMENTO, SI ES NEGATIVO, SE RESTA.
  def modificar_saldo(cantidad)  
    if(cantidad > 0)
      @saldo = @saldo + cantidad
    else
      @saldo = @saldo - cantidad
    end
  end
  
  #Devuelve el capital del que dispone el jugador, que es igual a su saldo más la suma de los valores de todas sus propiedades. El valor de una propiedad es la suma de su coste más el número de casas y hoteles que haya construidos por el precio de edificación. Si la propiedad estuviese hipotecada, se le restará el valor de la hipoteca base.
  def obtener_capital
    
  end
  
  #DEVUELTE LOS TITULOS PROPIEDAD DE JUGADORACTUAL
  def obtener_propiedades_hipotecadas(hipotecada)
    hipotecadas = Array.new
    @propiedades.each do |p|
      if(p.hipotecada == hipotecada)
        hipotecadas << p
      end
    end
    hipotecadas 
  end
  
  def pagar_cobrar_por_casa_y_hotel(cantidad)
    
  end
  
  def pagar_libertad(cantidad)
    
  end
  
  def puedo_edificar_casa(casilla)
    
  end
  
  def puedo_edificar_hotel(casilla)
    
  end
  
  def puedo_hipotecar(casilla)
    
  end

  def puedo_pagar_hipoteca(casilla)
    
  end
  
  #Cierto sólo si la casilla es de la propiedad de ese jugador y no la tiene hipotecada.
  def puedo_vender_propiedad(casilla)
    pruedo_vender = false
    if(es_de_mi_propiedad(casilla) == true && casilla.esta_hipotecada == false)
      puedo_vender = true
    end
    puedo_vender
  end
  
  #CIERTO SOLO SI LA CARTALIBERTAD NO ES NULA
  def tengo_carta_libertad
    libertad = false
    
    if(@carta_libertad != nil)
      libertad = true
    end
    
    libertad
  end
  
  def vender_propiedades(casilla)
    
  end

  
  #METODOS PRIVADOS
  
  private
  #DEVUELVE EL TOTAL DE CASA Y HOTELES QUE TIENE ESE JUGADOR EN SUS PROPIEDADES
  def cuantas_casas_hoteles_tengo()
    total = 0
    
    @propiedades.each do |p|
      total = total + (p.casilla.numHoteles + p.casilla.numCasas)
    end
    
    total
    
  end
  
  #ELIMINA EL TITULO PROPIEDAD DE ESA CASILLA DE SU LISTA DE PROPIEDADES
  def eliminar_de_mis_propiedades(casilla) 
    @propiedades.each do |p|
      if(p == casilla.tipo)
        @propiedades.remove_at(index(p))
      end
    end
  end
  
  #CIERTO SI EL JUGADOR TIENE ENTRE SUS PROPIEDADES EL TITULO DE PROPIEDAD DE ESA CASILLA
  def es_de_mi_propiedad(casilla)
    es = false
    @propiedades.each do |p|
      if(p.casilla == casilla)
        es = true
      end
    es
    end
  end
  
  def tengo_saldo(cantidad)
    tengo = false
    if(@saldo >= cantidad)
      tengo = true
    end
    tengo
  end
  
end #clase

end#modulo
